internal/contour/snapshot.go

```diff
@@ -0,0 +1,79 @@
+// Copyright Project Contour Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package contour
+
+import (
+	"github.com/envoyproxy/go-control-plane/pkg/cache/v2"
+	resource "github.com/envoyproxy/go-control-plane/pkg/resource/v2"
+	"github.com/projectcontour/contour/internal/dag"
+	contourxds "github.com/projectcontour/contour/internal/xds"
+	"github.com/sirupsen/logrus"
+)
+
+// SnapshotHandler implements the xDS snapshot cache
+type SnapshotHandler struct {
+	clusterCache   contourxds.Resource
+	listenerCache  contourxds.Resource
+	secretCache    contourxds.Resource
+	endpointsCache contourxds.Resource
+	routeCache     contourxds.Resource
+
+	// snapshotCache is a snapshot-based cache that maintains a single versioned
+	// snapshot of responses for xDS resources that Contour manages
+	snapshotCache cache.SnapshotCache
+
+	logrus.FieldLogger
+}
+
+// NewSnapshotHandler returns an instance of SnapShotHandler
+func NewSnapshotHandler(c cache.SnapshotCache, resources []ResourceCache, logger logrus.FieldLogger) *SnapshotHandler {
+
+	var clusterCache contourxds.Resource
+	var listenerCache contourxds.Resource
+	var secretCache contourxds.Resource
+	var endpointsCache contourxds.Resource
+	var routeCache contourxds.Resource
+
+	// Loop through resources to pick off the caches
+	// which will be used to create new snapshots.
+	for _, r := range resources {
+		switch r.TypeURL() {
+		case resource.ClusterType:
+			clusterCache = r
+		case resource.RouteType:
+			routeCache = r
+		case resource.ListenerType:
+			listenerCache = r
+		case resource.SecretType:
+			secretCache = r
+		case resource.EndpointType:
+			endpointsCache = r
+		}
+	}
+
+	return &SnapshotHandler{
+		snapshotCache:  c,
+		clusterCache:   clusterCache,
+		secretCache:    secretCache,
+		listenerCache:  listenerCache,
+		endpointsCache: endpointsCache,
+		routeCache:     routeCache,
+		FieldLogger:    logger,
+	}
```

Comment:
Nit:

What I had in mind was something along the lines of the pseudocode below. We can treat the resources generically, and control how we forward the DAG changes.


```Go
type SnapshotHandler struct {
    snapshot types.Snapshot
    resources []ResourceCache
}

func NewSnapshotHandler(c cache.SnapshotCache, resources []ResourceCache, logger logrus.FieldLogger) *SnapshotHandler {
}

func (s *SnapshotHandler) OnChange(root *dag.DAG) {
    // Forward the DAG change to each resource cache.
    for _, r := range s.resources {
	r.OnChange(root)
    }

    for _, r := range s.resources {
	contents := r.Contents()
	responseType := responseTypeOf(r.TypeURL())
	version := nextVersionForType()

	s.snapshot.Resources[responseType] = cache.NewResources(version, responseSlice(contents)
    }
}
```